src/ess/reduce/normalization.py

```diff
@@ -0,0 +1,244 @@
+    # that contain any unmasked data.
+    masks = {
+        name: mask.all(set(mask.dims) - {dim}) for name, mask in detector.masks.items()
+    }
+
+    # Prefer a bin coord over an event coord because this makes the behavior for binned
+    # and histogrammed data consistent. If we used an event coord, we might allow a
+    # monitor range that is less than the detector bins which is fine for the events,
+    # but would be wrong if the detector was subsequently histogrammed.
+    if (det_coord := detector.coords.get(dim)) is not None:
+        lo = sc.DataArray(det_coord[dim, :-1], masks=masks).nanmin().data
+        hi = sc.DataArray(det_coord[dim, 1:], masks=masks).nanmax().data
+    elif (det_coord := detector.bins.coords.get(dim)) is not None:
+        lo = sc.DataArray(det_coord, masks=masks).nanmin().data
+        hi = sc.DataArray(det_coord, masks=masks).nanmax().data
+    else:
+        raise sc.CoordError(
+            f"Missing '{dim}' coordinate in detector for monitor normalization."
+        )
+
+    if monitor.coords[dim].min() > lo or monitor.coords[dim].max() < hi:
+        raise ValueError(
+            f"Cannot normalize by monitor: The {dim} range of the monitor "
+            f"({monitor.coords[dim].min():c} to {monitor.coords[dim].max():c}) "
+            f"is smaller than the range of the detector ({lo:c} to {hi:c})."
+        )
+
+    if detector.bins is None:
+        # If we didn't rebin to the detector coord here, then, for a finer monitor
+        # binning than detector, the lookup table would extract one monitor value for
+        # each detector bin and ignore other values lying in the same detector bin.
+        # But integration would pick up all monitor bins.
+        return monitor.rebin({dim: det_coord})
+    return monitor[dim, lo:hi]
+
+
+def _mask_detector_for_norm(
+    *, detector: sc.DataArray, monitor: sc.DataArray
+) -> sc.DataArray:
+    """Mask the detector where the monitor is masked.
+
+    For performance, this applies the monitor mask to the detector bins.
+    This can lead to masking more vents than strictly necessary if we
+    used an event mask.
+    """
+    if (monitor_mask := _monitor_mask(monitor)) is None:
+        return detector
+
+    # Use rebin to reshape the mask to the detector:
+    dim = monitor.dim
+    mask = sc.DataArray(monitor_mask, coords={dim: monitor.coords[dim]}).rebin(
+        {dim: detector.coords[dim]}
+    ).data != sc.scalar(0, unit=None)
+    return detector.assign_masks({"_monitor_mask": mask})
+
+
+def _monitor_mask(monitor: sc.DataArray) -> sc.Variable | None:
+    """Mask nonfinite monitor values and combine all masks."""
+    masks = monitor.masks.values()
+
+    finite = sc.isfinite(monitor.data)
+    if not finite.all():
+        masks = itertools.chain(masks, (~finite,))
+
+    mask = None
+    for m in masks:
+        if mask is None:
+            mask = m
+        else:
+            mask |= m
+
+    return mask
```

Comment:
Reminds my of https://github.com/scipp/scipp/blob/d1e45c4d3476ebf6ffb41088696a99a53b6217d9/src/scipp/core/concepts.py#L76 but I don't know if that is any easier.